sdk/program/src/lib.rs

```diff
@@ -1,3 +1,650 @@
+//!
+//! [pdas]: https://docs.solana.com/developing/programming-model/calling-between-programs#program-derived-addresses
+//! [`Pubkey::find_program_address`]: pubkey::Pubkey::find_program_address
+//!
+//! A simple example of creating an account for a PDA:
+//!
+//! ```
+//! use solana_program::{
+//!     account_info::{next_account_info, AccountInfo},
+//!     entrypoint,
+//!     entrypoint::ProgramResult,
+//!     program::invoke_signed,
+//!     pubkey::Pubkey,
+//!     system_instruction,
+//!     system_program,
+//! };
+//!
+//! entrypoint!(process_instruction);
+//!
+//! fn process_instruction(
+//!     program_id: &Pubkey,
+//!     accounts: &[AccountInfo],
+//!     instruction_data: &[u8],
+//! ) -> ProgramResult {
+//!     let account_info_iter = &mut accounts.iter();
+//!     let payer = next_account_info(account_info_iter)?;
+//!     let vault_pda = next_account_info(account_info_iter)?;
+//!     let system_program = next_account_info(account_info_iter)?;
+//!
+//!     assert!(payer.is_writable);
+//!     assert!(payer.is_signer);
+//!     assert!(vault_pda.is_writable);
+//!     assert_eq!(vault_pda.owner, &system_program::ID);
+//!     assert!(system_program::check_id(system_program.key));
+//!
+//!     let vault_bump_seed = instruction_data[0];
+//!     let vault_seeds = &[b"vault", payer.key.as_ref(), &[vault_bump_seed]];
+//!     let expected_vault_pda = Pubkey::create_program_address(vault_seeds, program_id)?;
+//!
+//!     assert_eq!(vault_pda.key, &expected_vault_pda);
+//!
+//!     let lamports = 10000000;
+//!     let vault_size = 16;
+//!
+//!     invoke_signed(
+//!         &system_instruction::create_account(
+//!             &payer.key,
+//!             &vault_pda.key,
+//!             lamports,
+//!             vault_size,
+//!             &program_id,
+//!         ),
+//!         &[
+//!             payer.clone(),
+//!             vault_pda.clone(),
+//!         ],
+//!         &[
+//!             &[
+//!                 b"vault",
+//!                 payer.key.as_ref(),
+//!                 &[vault_bump_seed],
+//!             ],
+//!         ]
+//!     )?;
+//!     Ok(())
+//! }
+//! ```
+//!
+//! # Native programs
+//!
+//! Some solana programs are [_native programs_][np2], running native machine
+//! code that is distributed with the runtime, with well-known program IDs.
+//!
+//! [np2]: https://docs.solana.com/developing/runtime-facilities/programs
+//!
+//! Native programs are divided into several categories:
+//!
+//! __Built-ins__ act just like other Solana programs, in that they can
+//! be [invoked] from within Solana programs or from within [transactions]
+//! submitted by clients. They implement fundamental operations that require
+//! direct access to the runtime.
+//!
+//! [invoked]: #cross-program-instruction-execution
+//! [transactions]: https://docs.rs/solana-sdk/latest/solana_sdk/transaction/index.html
+//!
+//! __Special built-ins__ include a few programs that are defined like built-ins,
```

Comment:
What do you consider in this category?

Reply:
The compute budget program and the zk token proof program. They are the only built-ins I found that don't fall into the precompile or program loader category and also don't work with on-chain programs.

Reading through all the native programs and rationalizing for myself the differences between them was challenging. I'm happy for any suggestions about how to document them and distinguish their differences. I'm happy to rename this category that I made up, or document it in a different way.

Reply:
From my reading, these native programs could be invoked from other programs but their instructions are defined elsewhere:

- solana_vote_program
- solana_config_program
- address_lookup_table_program